torch_xla/csrc/runtime/pjrt_computation_client.h

```diff
@@ -231,10 +231,16 @@ class PjRtComputationClient : public ComputationClient {
                     std::vector<std::string> devices,
                     std::unique_ptr<xla::PjRtLoadedExecutable> executable)
         : Computation(std::move(computation), std::move(devices)),
-          executable(std::move(executable)) {}
+          executable(std::move(executable)) {
```

Comment:
If it's already being done as part of data placeholder prep, then this is redundant. It looks like we have to extract the output shardings as part of the `pjrt_computation` iniailzation, I think with AOT compilation and auto-sharding, it makes sense to do it strictly around the compilation (we will have to do it anyway there), and avoid doing it part of the execution call. Preferably, keep the `pjrt_copmputation` just a wrapper without much processing/overhead.

Reply:
Yeah, this is redundant. I only added this here because I wanted to avoid returning `UNKNOWN` out of `ExecuteReplicated`, and `GetOutputSharding` is far too slow to call per execution. This does run once per compilation and is cached. `PjRtComputation` is still a wrapper; it's just caching a really slow call we know we'll need down the line. We were effectively doing the same thing by saving `ProgramShape` in `Computation` already.

Depending on how auto-sharding works, and if we add the output shardings to the top-level `Computation`, the placeholders shardings could easily instead use the compiled executable's output shardings from here rather than digging into the HLO proto manually: https://github.com/pytorch/xla/blob/b9475d9ec910dd38dc4298951b0ddcc887ab3bdd/torch_xla/csrc/xla_sharding_util.cpp#L618-L633